### PR TITLE
Cli backwards compatibility testing

### PR DESCRIPTION
**Pull request by roastario**

### CODE REVIEW

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                results += checkAllPositionalCharactersArePresent(oldCommand, newCommand)
+            }
+        }
+
+        return results
+    }
+}
+
+
+data class CliBackwardsCompatibilityValidationCheck(val message: String)
```
Comment (tlil) [inline testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: Use different data classes per case perhaps? And encode the context in each instance. E.g., 

```kt
ParameterRemovedFromSubcommand(
    val subcommand: String, val parameter: String
)

ParameterTypeChange(
    val subcommand: String, val parameter: String, val oldType: Class<*>, val newType: Class<*>
)

// ...
```

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        return objectMapper.readValue<List<CommandDescription>>(inputStream, object : TypeReference<List<CommandDescription>>() {});
+    }
+
+    fun checkAllCommandsArePresent(old: List<CommandDescription>, new: List<CommandDescription>): List<CliBackwardsCompatibilityValidationCheck> {
+        val oldSet = old.map { it.commandName }.toSet()
+        val newSet = new.map { it.commandName }.toSet()
+        val newIsSuperSetOfOld = newSet.containsAll(oldSet)
+        return if (!newIsSuperSetOfOld) {
+            oldSet.filterNot { newSet.contains(it) }.map {
+                CliBackwardsCompatibilityValidationCheck("SubCommand: $it has been removed from the CLI interface")
```
Comment (tlil) [inline testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: CLI interface = command line interface interface, no? #nitpick

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    }
+
+    fun checkAllParamsAreOfTheSameType(old: CommandDescription, new: CommandDescription): List<CliBackwardsCompatibilityValidationCheck> {
+
+        val oldMap = old.params.map { it.parameterName to it.parameterType }.toMap()
+        val newMap = new.params.map { it.parameterName to it.parameterType }.toMap()
+
+        return oldMap.filter { newMap[it.key] != null && newMap[it.key] != it.value }.map {
+            CliBackwardsCompatibilityValidationCheck("Parameter [ ${it.key} has changed from type: ${it.value} to ${newMap[it.key]}")
```
Comment (tlil) [inline testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: For enum types, I wonder if it'd be helpful to traverse the possible enum values as well and make sure they don't introduce a breaking change. Wdyt?

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    private fun buildDescriptors(result: MutableList<CommandLine>): List<CommandDescription> {
+        return result.map { ::parseToDescription.invoke(it) }
+    }
+
+    internal fun parseToDescription(it: CommandLine): CommandDescription {
+        val commandSpec = it.commandSpec
+        val options = commandSpec.options().filterNot { it.usageHelp() || it.versionHelp() }
+                .map { hit -> hit.names().map { it to hit } }
```
Comment (tlil) [inline testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: Can combine the `map` and `flatMap`

File: testing/test-cli/src/test/kotlin/net/corda/testing/DummyCliCompatibilityTest.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    var depth: Int? = null
+
+    @CommandLine.Parameters(index = "0")
+    var machine: String? = null
+    @CommandLine.Parameters(index = "1")
+    var listStyle: Int = 0
+}
+
+class DummyCliCompatibilityTest : CliBackwardsCompatibleTest() {
+    @org.junit.Test
```
Comment (tlil) [inline testing/test-cli/src/test/kotlin/net/corda/testing/DummyCliCompatibilityTest.kt]: `@Test` ? #nitpick

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        val resourceAsStream = this.javaClass.classLoader.getResourceAsStream("$commandLineToCheckName.yml")
+                ?: throw IllegalStateException("no Descriptor for $commandLineToCheckName found on classpath")
+        val old = readCommandDescription(resourceAsStream)
+        val new = topoSort(CommandLine(instance))
+        val results = ArrayList<CliBackwardsCompatibilityValidationCheck>()
+        results += checkAllCommandsArePresent(old, new)
+        for (oldCommand in old) {
+            new.find { it.commandName == oldCommand.commandName }?.let { newCommand ->
+                results += checkAllOptionsArePresent(oldCommand, newCommand)
+                results += checkAllParamsAreOfTheSameType(oldCommand, newCommand)
```
Comment (tlil) [inline testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: What's your take on checking default values, by the way? You can glean insight into what the default values are from inspecting the command instances.

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt (hunk 4 above)
Comment (roastario) [reply testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: I know it adds an extra line, but I *personally* find it easier to understand when these stream operations are not collapsed.

ie;
.filter{ it.blah == foo}.first() 
vs
first{it.blah == foo}

makes it easier to understand for me, and there is zero overhead computationally.

Comment (tlil) [reply testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: Fair 'nuf

File: build.gradle [hunk 7]
```diff
@@ -171,7 +172,7 @@ allprojects {
             jvmTarget = "1.8"
             javaParameters = true   // Useful for reflection.
             freeCompilerArgs = ['-Xjvm-default=compatibility']
-            allWarningsAsErrors = project.hasProperty('compilation.allWarningsAsErrors') ? project.property('compilation.allWarningsAsErrors').toBoolean() : true
+            allWarningsAsErrors = false
```
Comment (tlil) [inline build.gradle]: I'm guessing this wasn't supposed to make it in :)

File: testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt (hunk 6 above)
Comment (tlil) [reply testing/test-cli/src/main/kotlin/net/corda/testing/GenerateCommandLineCompat.kt]: Can leave this for later

**Pull request merged**